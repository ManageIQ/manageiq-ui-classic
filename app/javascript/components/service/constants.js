export const DIALOG_FIELD_TYPES = {
  checkBox: 'DialogFieldCheckBox',
  date: 'DialogFieldDateControl',
  dateTime: 'DialogFieldDateTimeControl',
  dropDown: 'DialogFieldDropDownList',
  radio: 'DialogFieldRadioButton',
  tag: 'DialogFieldTagControl',
  textBox: 'DialogFieldTextBox',
  textArea: 'DialogFieldTextAreaBox',
};

export const ServiceType = {
  order: 'orderServiceForm',
  dialog: 'serviceDialog',
  request: 'serviceRequest',
  reconfigure: 'serviceReconfigure',
};

export const RefreshStatus = {
  notStarted: 'notStarted',
  completed: 'completed',
};
