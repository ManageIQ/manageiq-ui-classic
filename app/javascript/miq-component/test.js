import miqComponent from './index';

import {
  getDefinition,
  sanitizeAndFreezeInstanceId,
  validateInstance,
  define,
  newInstance,
  getInstance,
  isDefined,
  getComponentNames,
  getComponentInstances,
  clearRegistry
} from './registry';

import {
  writeProxy,
  lockInstanceProperties
} from './utils';

describe('Component API', () => {

  beforeAll(() => {
    miqComponent();
  });

  afterEach(() => {
    clearRegistry();
  });

  it('ManageIQ.component namespace is defined', () => {
    expect(ManageIQ.component).toBeDefined();
    expect(ManageIQ.component.define).toEqual(expect.any(Function));
    expect(ManageIQ.component.newInstance).toEqual(expect.any(Function));
    expect(ManageIQ.component.getInstance).toEqual(expect.any(Function));
    expect(ManageIQ.component.isDefined).toEqual(expect.any(Function));
  });

  it('define method can be used to register new components', () => {
    define('FooComponent', {});
    define('BarComponent', {}, [{}]);
    expect(isDefined('FooComponent')).toBe(true);
    expect(isDefined('BarComponent')).toBe(true);
    expect(getComponentNames()).toEqual(['FooComponent', 'BarComponent']);
  });

  it('define method does nothing if the component name is already taken', () => {
    define('FooComponent', {});
    define('FooComponent', {});
    expect(getComponentNames()).toEqual(['FooComponent']);
  });

  it('define method does nothing if the component name is not a string', () => {
    define(123, {});
    expect(isDefined(123)).toBe(false);
    expect(getComponentNames()).toEqual([]);
  });

  it('define method can be used associate existing instances with the new component', () => {
    var testInstances = [
      { id: 'first' }, { id: 'second' }
    ];

    define('FooComponent', {}, testInstances);
    expect(getInstance('FooComponent', 'first')).toBe(testInstances[0]);
    expect(getInstance('FooComponent', 'second')).toBe(testInstances[1]);
  });

  it('when passing existing instances, define method ensures a sane instance id', () => {
    var testInstances = [
      { id: 'first' }, { id: 123 }, {}
    ];

    define('FooComponent', {}, testInstances);

    var registeredInstances = getComponentInstances('FooComponent');
    expect(registeredInstances.length).toBe(3);
    expect(registeredInstances[0].id).toBe('first');
    expect(registeredInstances[1].id).toBe('FooComponent-1');
    expect(registeredInstances[2].id).toBe('FooComponent-2');
  });

  it('when passing existing instances, define method ensures that instance id is frozen', () => {
    var testInstances = [
      { id: 'first' }
    ];

    define('FooComponent', {}, testInstances);
    expect(() => {
      testInstances[0].id = 'second';
    }).toThrow();
  });

  it('when passing existing instances, define method skips falsy values', () => {
    var testInstances = [
      false, '', null, undefined, {}
    ];

    define('FooComponent', {}, testInstances);

    var registeredInstances = getComponentInstances('FooComponent');
    expect(registeredInstances.length).toBe(1);
    expect(registeredInstances[0].id).toBe('FooComponent-0');
  });

  it('when passing existing instances, define method throws in case of reference duplicity', () => {
    var testInstance = { id: 'first' };

    expect(() => {
      define('FooComponent', {}, [testInstance, testInstance]);
    }).toThrow();
  });

  it('when passing existing instances, define method throws in case of id duplicity', () => {
    var testInstances = [
      { id: 'first' }, { id: 'first' }
    ];

    expect(() => {
      define('FooComponent', {}, testInstances);
    }).toThrow();
  });

  it('newInstance method can be used to create new component instances', () => {
    var testInstances = [
      { id: 'first' }, { id: 'second' }
    ];

    var testBlueprint = {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => testInstances[0])
                .mockImplementationOnce(() => testInstances[1])
    };

    define('FooComponent', testBlueprint);

    var mountFirstInstanceTo = document.createElement('div');
    var resultingInstances = [
      newInstance('FooComponent', { bar: 123 }, mountFirstInstanceTo),
      newInstance('FooComponent', { baz: true })
    ];

    expect(testBlueprint.create).toHaveBeenCalledTimes(2);
    expect(testBlueprint.create.mock.calls[0]).toEqual([
      { bar: 123 },
      mountFirstInstanceTo
    ]);
    expect(testBlueprint.create.mock.calls[1]).toEqual([
      { baz: true },
      undefined
    ]);

    expect(resultingInstances[0]).toBe(testInstances[0]);
    expect(resultingInstances[1]).toBe(testInstances[1]);

    expect(resultingInstances[0].id).toBe('first');
    expect(resultingInstances[1].id).toBe('second');

    expect(resultingInstances[0].props).toEqual({ bar: 123 });
    expect(resultingInstances[1].props).toEqual({ baz: true });

    resultingInstances.forEach(instance => {
      expect(instance.update).toEqual(expect.any(Function));
      expect(instance.destroy).toEqual(expect.any(Function));
    });

    var registeredInstances = getComponentInstances('FooComponent');
    expect(registeredInstances.length).toBe(2);
    expect(registeredInstances[0]).toBe(resultingInstances[0]);
    expect(registeredInstances[1]).toBe(resultingInstances[1]);
  });

  it('newInstance method does nothing if the component is not already defined', () => {
    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    expect(resultingInstance).toBeUndefined();
  });

  it('newInstance method does nothing if blueprint.create is not a function', () => {
    define('FooComponent', {});

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    expect(resultingInstance).toBeUndefined();
  });

  it('newInstance method throws if blueprint.create returns a falsy value', () => {
    define('FooComponent', {
      create() { return null; }
    });

    expect(() => {
      newInstance('FooComponent', { bar: 123 });
    }).toThrow();
  });

  it('newInstance method ensures a sane instance id', () => {
    define('FooComponent', {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => ({ id: 'first' }))
                .mockImplementationOnce(() => ({ id: 123 }))
                .mockImplementationOnce(() => ({}))
    });

    var resultingInstances = [
      newInstance('FooComponent', { bar: 123 }),
      newInstance('FooComponent', { baz: true }),
      newInstance('FooComponent', { qux: ['1337'] })
    ];

    expect(resultingInstances[0].id).toBe('first');
    expect(resultingInstances[1].id).toBe('FooComponent-1');
    expect(resultingInstances[2].id).toBe('FooComponent-2');
  });

  it('newInstance method ensures that instance id is frozen', () => {
    define('FooComponent', {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => ({ id: 'first' }))
    });

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    expect(() => {
      resultingInstance.id = 'second';
    }).toThrow();
  });

  it('newInstance method throws if blueprint.create returns the same instance', () => {
    var testInstance = { id: 'first' };

    define('FooComponent', {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => testInstance)
                .mockImplementationOnce(() => testInstance)
    });

    newInstance('FooComponent', { bar: 123 });

    expect(() => {
      newInstance('FooComponent', { bar: 123 });
    }).toThrow();
  });

  it('newInstance method throws if blueprint.create returns an instance with id already taken', () => {
    define('FooComponent', {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => ({ id: 'first' }))
                .mockImplementationOnce(() => ({ id: 'first' }))
    });

    newInstance('FooComponent', { bar: 123 });

    expect(() => {
      newInstance('FooComponent', { bar: 123 });
    }).toThrow();
  });

  it('trying to rewrite instance props throws an error', () => {
    define('FooComponent', {
      create() { return {}; }
    });

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    expect(() => {
      resultingInstance.props = {};
    }).toThrow();
  });

  it('instance.update merges props and delegates to blueprint.update', () => {
    var testBlueprint = {
      create() { return {}; },
      update: jest.fn().mockName('testBlueprint.update')
    };

    define('FooComponent', testBlueprint);

    var mountFirstInstanceTo = document.createElement('div');
    var resultingInstances = [
      newInstance('FooComponent', { bar: 123 }, mountFirstInstanceTo),
      newInstance('FooComponent', { baz: true })
    ];

    resultingInstances[0].update({ baz: true, qux: ['1337'] });
    resultingInstances[1].update({ baz: false });

    expect(testBlueprint.update).toHaveBeenCalledTimes(2);
    expect(testBlueprint.update.mock.calls[0]).toEqual([
      { bar: 123, baz: true, qux: ['1337'] },
      mountFirstInstanceTo
    ]);
    expect(testBlueprint.update.mock.calls[1]).toEqual([
      { baz: false },
      undefined
    ]);

    expect(resultingInstances[0].props).toEqual({
      bar: 123, baz: true, qux: ['1337']
    });
    expect(resultingInstances[1].props).toEqual({
      baz: false
    });
  });

  it('instance.update does nothing if blueprint.update is not a function', () => {
    define('FooComponent', {
      create() { return {}; }
    });

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    resultingInstance.update({ baz: true, qux: ['1337'] });

    expect(resultingInstance.props).toEqual({ bar: 123 });
  });

  it('multiple props modifications will trigger single instance.update', (done) => {
    var testBlueprint = {
      create() { return {}; },
      update: jest.fn().mockName('testBlueprint.update')
    };

    define('FooComponent', testBlueprint);

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    resultingInstance.update = jest.fn().mockName('resultingInstance.update');

    resultingInstance.props.baz = true;
    resultingInstance.props.qux = ['1337'];

    setTimeout(() => {
      expect(resultingInstance.update).toHaveBeenCalledWith(
        { baz: true, qux: ['1337'] }
      );
      done();
    }, 0);
  });

  it('instance.destroy delegates to blueprint.destroy', () => {
    var testBlueprint = {
      create() { return {}; },
      destroy: jest.fn().mockName('testBlueprint.destroy')
    };

    define('FooComponent', testBlueprint);

    var mountFirstInstanceTo = document.createElement('div');
    var resultingInstances = [
      newInstance('FooComponent', { bar: 123 }, mountFirstInstanceTo),
      newInstance('FooComponent', { baz: true })
    ];

    resultingInstances[0].destroy();
    resultingInstances[1].destroy();

    expect(testBlueprint.destroy).toHaveBeenCalledTimes(2);
    expect(testBlueprint.destroy.mock.calls[0]).toEqual([
      resultingInstances[0],
      mountFirstInstanceTo
    ]);
    expect(testBlueprint.destroy.mock.calls[0][0]).toBe(resultingInstances[0]);
    expect(testBlueprint.destroy.mock.calls[1]).toEqual([
      resultingInstances[1],
      undefined
    ]);
    expect(testBlueprint.destroy.mock.calls[1][0]).toBe(resultingInstances[1]);
  });

  it('instance.destroy removes the component instance', () => {
    define('FooComponent', {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => ({ id: 'first' }))
    });

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    resultingInstance.destroy();

    expect(getInstance('FooComponent', 'first')).toBeUndefined();
    expect(getComponentInstances('FooComponent')).toEqual([]);
  });

  it('instance.destroy prevents access to existing instance properties except for id', () => {
    define('FooComponent', {
      create() { return {}; }
    });

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    resultingInstance.destroy();

    Object.keys(resultingInstance)
      .filter(propName => propName !== 'id')
      .forEach(propName => {
        expect(() => {
          resultingInstance[propName];
        }).toThrow();
        expect(() => {
          resultingInstance[propName] = ['1337'];
        }).toThrow();
      });
  });

  it('getInstance method can be used to obtain existing component instances', () => {
    define('FooComponent', {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => ({ id: 'first' }))
    });

    var resultingInstance = newInstance('FooComponent', { bar: 123 });
    var obtainedInstance = getInstance('FooComponent', 'first');
    expect(obtainedInstance).toBe(resultingInstance);
    expect(getInstance('FooComponent', 'second')).toBeUndefined();
  });

  it('isDefined method can be used to determine whether a component is already defined', () => {
    define('FooComponent', {});
    expect(isDefined('FooComponent')).toBe(true);
    expect(isDefined('BarComponent')).toBe(false);
  });

  it('getDefinition returns the correct component definition', () => {
    var testBlueprint = {};
    define('FooComponent', testBlueprint);

    var definition = getDefinition('FooComponent');
    expect(definition.name).toBe('FooComponent');
    expect(definition.blueprint).toBe(testBlueprint);

    expect(getDefinition('BarComponent')).toBeUndefined();
  });

  it('sanitizeAndFreezeInstanceId ensures the instance id is sane and cannot be changed', () => {
    var testInstances = [
      { id: 'first' }, { id: 123 }, {}
    ];

    define('FooComponent', {});
    var definition = getDefinition('FooComponent');

    testInstances.forEach(instance => {
      sanitizeAndFreezeInstanceId(instance, definition);
    });

    expect(testInstances[0].id).toBe('first');
    expect(testInstances[1].id).toBe('FooComponent-0');
    expect(testInstances[2].id).toBe('FooComponent-0');

    testInstances.forEach(instance => {
      expect(() => {
        instance.id = 'second';
      }).toThrow();
    });
  });

  it('validateInstance performs the necessary instance validations', () => {
    define('FooComponent', {
      create: jest.fn().mockName('testBlueprint.create')
                .mockImplementationOnce(() => ({ id: 'first' }))
    });

    var firstInstance = newInstance('FooComponent', { bar: 123 });
    var definition = getDefinition('FooComponent');

    expect(() => {
      validateInstance({ id: 'second' }, definition);
    }).not.toThrow();

    expect(() => {
      validateInstance(firstInstance, definition);
    }).toThrow();

    expect(() => {
      validateInstance({ id: 'first' }, definition);
    }).toThrow();
  });

  it('writeProxy can be used to proxy writes to properties of the given object', () => {
    var object = { bar: 123, baz: true };
    var onWrite = jest.fn().mockName('onWrite');

    var resultingObject = writeProxy(object, onWrite);
    expect(resultingObject).toEqual(object);

    resultingObject.bar = 456;
    resultingObject.qux = ['1337'];

    expect(resultingObject).toEqual(object);
    expect(object).toEqual({ bar: 456, baz: true, qux: ['1337'] });

    expect(onWrite).toHaveBeenCalledTimes(2);
    expect(onWrite.mock.calls[0]).toEqual([
      'bar', 456
    ]);
    expect(onWrite.mock.calls[1]).toEqual([
      'qux', ['1337']
    ]);
  });

  it('lockInstanceProperties prevents access to existing instance properties except for id', () => {
    var testInstance = { id: 'first', bar: 123, baz: true };

    var resultingInstance = lockInstanceProperties(testInstance);
    expect(resultingInstance.id).toBe('first');

    ['bar', 'baz'].forEach(propName => {
      expect(() => {
        resultingInstance[propName];
      }).toThrow();
      expect(() => {
        resultingInstance[propName] = ['1337'];
      }).toThrow();
    });
  });

});
