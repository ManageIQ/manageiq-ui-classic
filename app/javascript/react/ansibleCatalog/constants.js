import PropTypes from 'prop-types';

export const DROPDOWN_OPTIONS = {
  LOGGING_OUTPUT: [
    { label: __('On Error'), value: 'on_error' },
    { label: __('Always'), value: 'always' },
    { label: __('Never'), value: 'never' },
  ],
  VERBOSITY: [
    { label: __('0 (Normal)'), value: 0 },
    { label: __('1 (Verbose)'), value: 1 },
    { label: __('2 (More Verbose)'), value: 2 },
    { label: __('3 (Debug)'), value: 3 },
    { label: __('4 (Connection Debug)'), value: 4 },
    { label: __('5 (WinRM Debug)'), value: 5 },
  ],
  DIALOG: [
    { label: __('Use Existing'), value: 'existing' },
    { label: __('Create New'), value: 'new' },
  ],
  PLAYBOOK: [
    { label: __('No'), value: 'no_with_playbook' },
    { label: __('Before Playbook runs'), value: 'pre_with_playbook' },
    { label: __('After Playbook runs'), value: 'post_with_playbook' },
  ],
};
export const FORM_DEFAULTS = {
  name: '',
  description: '',
  display: false,
  long_description: '',
  service_template_catalog_id: '',
  remove_resources: 'no_without_playbook',
  dialog_id: '',
  dialogOption: 'existing',
};

export const ANSIBLE_FIELDS = ['repository_id',
  'playbook_id', 'credential_id', 'vault_credential_id',
  'cloud_type', 'cloud_credential_id', 'execution_ttl', 'hosts',
  'become_enabled', 'verbosity', 'log_output', 'extra_vars'];

export const DEFAULT_PLACEHOLDER = __('<Choose>');

export const CATALOG_ITEM_PROPS = {
  ansibleCatalog: PropTypes.shape({}).isRequired,
  loadCatalogs: PropTypes.func.isRequired,
  loadDialogs: PropTypes.func.isRequired,
  loadRepos: PropTypes.func.isRequired,
  loadPlaybooks: PropTypes.func.isRequired,
  loadCloudCredential: PropTypes.func.isRequired,
  loadCredentials: PropTypes.func.isRequired,
  loadCloudTypes: PropTypes.func.isRequired,
  loadCatalogItem: PropTypes.func.isRequired,
  duplicateDropdowns: PropTypes.func.isRequired,
  region: PropTypes.number.isRequired,
  catalogItemFormId: PropTypes.number,
  availableCatalogs: PropTypes.array,
};

export const PROVISIONING_FORM_FIELDS = {
  repository_id: '',
  playbook_id: '',
  credential_id: '',
  vault_credential_id: '',
  cloud_type: '',
  cloud_credential_id: '',
  execution_ttl: '',
  hosts: 'localhost',
  become_enabled: false,
  verbosity: 0,
  log_output: 'on_error',
  extra_vars: [],
};
export const ANSIBLE_PLAYBOOK_FIELDS_PROPS = {
  region: PropTypes.number,
  prefix: PropTypes.string,
  loadPlaybooks: PropTypes.func.isRequired,
  loadCloudCredentials: PropTypes.func.isRequired,
  formValues: PropTypes.shape({}).isRequired,
  initialValues: PropTypes.shape({}).isRequired,
  changeField: PropTypes.func.isRequired,
  addExtraVars: PropTypes.func.isRequired,
  children: PropTypes.node,
};
