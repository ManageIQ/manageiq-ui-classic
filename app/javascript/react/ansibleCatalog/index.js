import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as catalogActions from './catalogActions';

import reducer from './catalogReducer';
import { AnsibleCatalogItemForm } from './ansibleCatalogItemForm';

const actions = { ...catalogActions };
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);
const mapStateToProps = state => ({
  ansibleCatalog: state.ansibleCatalog || {},
});
ManageIQ.redux.addReducer({ ansibleCatalog: reducer });
export default connect(mapStateToProps, mapDispatchToProps)(AnsibleCatalogItemForm);
