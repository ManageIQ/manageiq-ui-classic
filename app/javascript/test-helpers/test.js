import { spyOnDeeply } from './jest-mocks';

describe('Test helpers', () => {

  it('spyOnDeeply deep-clones mock function call arguments', () => {
    const object = {
      test(foo, bar) { return true; }
    };

    const mock = spyOnDeeply(object, 'test');
    expect(jest.isMockFunction(mock)).toBe(true);

    let fooArg = ['foo'];
    let barArg = { bar: true };

    expect(object.test(fooArg, barArg)).toBe(true);
    expect(mock).toHaveBeenCalledWith(
      ['foo'],
      { bar: true }
    );

    fooArg = 'foo';

    expect(object.test(fooArg, barArg)).toBe(true);
    expect(mock).toHaveBeenCalledWith(
      'foo',
      { bar: true }
    );
  });

});
