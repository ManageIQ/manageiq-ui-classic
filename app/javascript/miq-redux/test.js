import { spyOnDeeply } from '../test-helpers/jest-mocks'
import miqRedux from './index';

import {
  rootReducer,
  addReducer,
  applyReducerHash,
  clearReducers
} from './reducer';

describe('Redux API', () => {

  let $ngRedux;
  let testReducers;

  beforeAll(() => {
    miqRedux(ManageIQ.angular.app);
  });

  beforeEach(() => {
    angular.mock.module('ManageIQ');
    angular.mock.inject((_$ngRedux_) => {
      $ngRedux = _$ngRedux_;
    });
  });

  beforeEach(() => {
    testReducers = {

      init(state, action) {
        if (action.type === 'TEST_INIT') {
          return { foo: 'hello', bar: true };
        }
        return state;
      },

      foo(state, action) {
        if (action.type === 'SET_FOO') {
          return Object.assign(state, { foo: action.payload });
        }
        return state;
      },

      bar(state, action) {
        if (action.type === 'RESET_BAR') {
          return Object.assign(state, { bar: false });
        }
        return state;
      }

    };

    Object.keys(testReducers).forEach(reducerName => {
      const spy = spyOnDeeply(testReducers, reducerName).mockName(`${reducerName}Reducer`);
      testReducers[`${reducerName}Spy`] = spy;
    });
  });

  afterEach(() => {
    clearReducers();
  });

  it('ManageIQ.redux namespace is defined', () => {
    expect(ManageIQ.redux).toBeDefined();
    expect(ManageIQ.redux.store).toBe($ngRedux);
    expect(ManageIQ.redux.addReducer).toEqual(expect.any(Function));
    expect(ManageIQ.redux.applyReducerHash).toEqual(expect.any(Function));
  });

  it('$ngRedux contains common Redux store methods', () => {
    expect($ngRedux.getState).toEqual(expect.any(Function));
    expect($ngRedux.dispatch).toEqual(expect.any(Function));
    expect($ngRedux.subscribe).toEqual(expect.any(Function));
  });

  it('$ngRedux complies with the standard Redux workflow', () => {
    const storeSubscribe = jest.fn().mockName('storeSubscribe');

    $ngRedux.subscribe(storeSubscribe);

    addReducer(testReducers.init);
    addReducer(testReducers.foo);
    addReducer(testReducers.bar);

    $ngRedux.dispatch({ type: 'TEST_INIT' });
    expect($ngRedux.getState()).toEqual({ foo: 'hello', bar: true });

    $ngRedux.dispatch({ type: 'SET_FOO', payload: 123 });
    expect($ngRedux.getState()).toEqual({ foo: 123, bar: true });

    $ngRedux.dispatch({ type: 'RESET_BAR' });
    expect($ngRedux.getState()).toEqual({ foo: 123, bar: false });

    expect(testReducers.initSpy).toHaveBeenCalledTimes(3);
    expect(testReducers.fooSpy).toHaveBeenCalledTimes(3);
    expect(testReducers.barSpy).toHaveBeenCalledTimes(3);
    expect(storeSubscribe).toHaveBeenCalledTimes(3);
  });

  it('rootReducer invokes all reducers registered via addReducer', () => {
    addReducer(testReducers.foo);
    addReducer(testReducers.bar);

    const newState = rootReducer(
      { foo: 'hello', bar: true },
      { type: 'SET_FOO', payload: 123 }
    );
    expect(newState).toEqual({ foo: 123, bar: true });

    expect(testReducers.fooSpy).toHaveBeenCalledWith(
      { foo: 'hello', bar: true },
      { type: 'SET_FOO', payload: 123 }
    );
    expect(testReducers.barSpy).toHaveBeenCalledWith(
      { foo: 123, bar: true },
      { type: 'SET_FOO', payload: 123 }
    );
  });

  it('rootReducer returns original state for an unknown action', () => {
    addReducer(testReducers.foo);
    addReducer(testReducers.bar);

    const originalState = { foo: 'hello', bar: true };
    const newState = rootReducer(
      originalState,
      { type: 'UNKNOWN' }
    );
    expect(newState).toBe(originalState);

    [testReducers.fooSpy, testReducers.barSpy].forEach(spy => {
      expect(spy).toHaveBeenCalledWith(
        { foo: 'hello', bar: true },
        { type: 'UNKNOWN' }
      );
    });
  });

  it('addReducer returns function to remove that reducer', () => {
    addReducer(testReducers.init);
    const removeFooReducer = addReducer(testReducers.foo);

    $ngRedux.dispatch({ type: 'TEST_INIT' });
    removeFooReducer();
    $ngRedux.dispatch({ type: 'TEST_INIT' });

    expect(testReducers.initSpy).toHaveBeenCalledTimes(2);
    expect(testReducers.fooSpy).toHaveBeenCalledTimes(1);
  });

  it('applyReducerHash invokes the right reducer if present', () => {
    const reducerHash = {
      'TEST_INIT': testReducers.init,
      'SET_FOO': testReducers.foo,
      'RESET_BAR': testReducers.bar
    };

    const newState = applyReducerHash(
      reducerHash,
      { foo: 'hello', bar: true },
      { type: 'SET_FOO', payload: 123 }
    );
    expect(newState).toEqual({
      foo: 123,
      bar: true
    });

    expect(testReducers.initSpy).toHaveBeenCalledTimes(0);
    expect(testReducers.fooSpy).toHaveBeenCalledTimes(1);
    expect(testReducers.barSpy).toHaveBeenCalledTimes(0);
  });

});
