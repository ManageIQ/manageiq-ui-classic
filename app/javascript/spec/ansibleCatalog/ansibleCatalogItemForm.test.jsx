import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import PropTypes from 'prop-types';
import fetchMock from 'fetch-mock';
import { reduxStore, sampleForm } from './test.fixtures';
import { ANSIBLE_FIELDS } from '../../react/ansibleCatalog/constants';
import AnsibleCatalogItemForm from '../../react/ansibleCatalog';
import { AnsibleCatalogItemForm as AnsibleCatalogItemForm1 } from '../../react/ansibleCatalog/ansibleCatalogItemForm';


describe('Ansible catalog item form component', () => {
  const mockStore = configureStore([thunk]);

  let store;
  let initialProps;
  const shallowMount = () => shallow(
    <AnsibleCatalogItemForm {...initialProps} />,
    {
      context: { store },
      childContextTypes: { store: PropTypes.object.isRequired },
    },
  ).dive();

  beforeEach(() => {
    store = mockStore({
      ansibleCatalog: {
        ...reduxStore,
      },
    });
    initialProps = {
      loadCatalogs: jest.fn(),
      loadDialogs: jest.fn(),
      loadRepos: jest.fn(),
      loadCatalogItem: jest.fn(),
      loadPlaybooks: jest.fn(),
      loadCredentials: jest.fn(),
      loadCloudTypes: jest.fn(),
      loadCloudCredential: jest.fn(),
      duplicateDropdowns: jest.fn(),
      region: 10,
    };
  });
  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });
  it('should render correctly', () => {
    const wrapper = shallowMount();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should allow a catalog item to be edited', () => {
    const catalogItemId = 1234;

    const loadCatalogItem = jest.fn(() => Promise.resolve({}));
    const loadCredential = jest.fn(() => Promise.resolve({}));
    mount(
      <AnsibleCatalogItemForm1
        {...initialProps}
        ansibleCatalog={{ ...reduxStore }}
        loadCatalogItem={loadCatalogItem}
        loadCloudCredential={loadCredential}
        catalogItemFormId={catalogItemId}
      />,
      {
        context: { store },
        childContextTypes: { store: PropTypes.object.isRequired },
      },
    );
    expect(loadCatalogItem).toHaveBeenCalledWith(1234);
  });
  it('should set default values when a catalog item is being edited', (done) => {
    const catalogItemId = 1234;

    store = mockStore({
      ansibleCatalog: {
        ...reduxStore,
      },
    });
    const loadCatalogItem = jest.fn(() => Promise.resolve({}));
    const loadCloudCredential = jest.fn(() => Promise.resolve({}));
    const loadPlaybooks = jest.fn(() => Promise.resolve({}));
    const props = {
      ...initialProps, loadCatalogItem, loadCloudCredential, loadPlaybooks,
    };
    const wrapper = mount(
      <AnsibleCatalogItemForm1
        {...props}
        ansibleCatalog={{ ...reduxStore }}
        catalogItemFormId={catalogItemId}
      />
      , {
        context: { store },
        childContextTypes: { store: PropTypes.object.isRequired },
      },
    );

    expect(loadCatalogItem).toHaveBeenCalledWith(1234);
    wrapper.update();
    setImmediate(() => {
      done();
      expect(loadCloudCredential).toHaveBeenCalledWith(123);
      const formValues = wrapper.state().initialValues;
      expect(formValues).toEqual(sampleForm);
    });
  });
  it('should allow a form to be submitted', () => {
    const wrapper = shallowMount();
    window.location.assign = jest.fn();
    fetchMock.postOnce('/api/service_templates/', {});
    const values = {
      name: 'test',
      description: '',
      long_description: '',
      service_template_catalog_id: '',
      provision_extra_vars: [],
      retirement_extra_vars: [],
    };
    wrapper.instance().submitForm(values);
    expect(fetchMock.called('/api/service_templates/')).toBe(true);
  });
  it('should report an error if submitting failed', () => {
    // eslint-disable-next-line no-console
    console.error = jest.fn();
    const spy = jest.spyOn(window, 'add_flash');
    window.add_flash = spy;
    const wrapper = shallowMount();
    fetchMock.mock('/api/service_templates/', {
      body: '',
      status: 500,
    });
    const values = {
      name: 'test',
      description: '',
      long_description: '',
      service_template_catalog_id: '',
      provision_extra_vars: [],
      retirement_extra_vars: [],
    };
    wrapper.instance().submitForm(values);
    wrapper.update();
    setImmediate(() => {
      expect(spy).toHaveBeenCalledWith('Catalog item failed to be added. ', 'error');
    });
  });
  it('should close a popup', () => {
    const wrapper = shallowMount();
    const instance = wrapper.instance();
    instance.setState({ showCopyProvisionDialog: true });
    instance.closeCopyProvisionDialog();
    expect(wrapper.state.showCopyProvisionDialog).toBeFalsy();
  });
  it('should show a message when canceling form', () => {
    const wrapper = shallowMount();
    const instance = wrapper.instance();
    const flashMessage = jest.fn();
    global.miqFlashLater = flashMessage;
    instance.handleCancel();
    expect(flashMessage).toHaveBeenCalledWith({
      message: 'Add of Catalog Item was cancelled by the user',
    });
  });
  it('should copy Form values', () => {
    const wrapper = shallowMount();
    const instance = wrapper.instance();
    const changeValue = jest.fn();
    const utils = {
      changeValue,
    };
    const formState = {
      formState:
      {
        values: sampleForm,
      },
    };
    instance.copyFormValues(null, formState, utils);
    expect(changeValue).toHaveBeenCalledTimes(ANSIBLE_FIELDS.length);
  });
});
