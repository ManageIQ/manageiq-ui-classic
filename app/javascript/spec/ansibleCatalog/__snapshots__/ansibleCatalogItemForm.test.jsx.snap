// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ansible catalog item form component should render correctly 1`] = `
<div>
  <div
    id="flash_msg_div"
  />
  <ReactFinalForm
    initialValues={
      Object {
        "description": "",
        "dialogOption": "existing",
        "dialog_id": "",
        "display": false,
        "long_description": "",
        "name": "",
        "provision_become_enabled": false,
        "provision_cloud_credential_id": "",
        "provision_cloud_type": "",
        "provision_credential_id": "",
        "provision_execution_ttl": "",
        "provision_extra_vars": Array [],
        "provision_hosts": "localhost",
        "provision_log_output": "on_error",
        "provision_playbook_id": "",
        "provision_repository_id": "",
        "provision_vault_credential_id": "",
        "provision_verbosity": 0,
        "remove_resources": "no_without_playbook",
        "retirement_become_enabled": false,
        "retirement_cloud_credential_id": "",
        "retirement_cloud_type": "",
        "retirement_credential_id": "",
        "retirement_execution_ttl": "",
        "retirement_extra_vars": Array [],
        "retirement_hosts": "localhost",
        "retirement_log_output": "on_error",
        "retirement_playbook_id": "",
        "retirement_repository_id": "",
        "retirement_vault_credential_id": "",
        "retirement_verbosity": 0,
        "service_template_catalog_id": "",
      }
    }
    mutators={
      Object {
        "copyValues": [Function],
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[Function]}
    render={[Function]}
  />
</div>
`;
